src/Raven.Server/ServerWide/ClusterStateMachine.cs

```diff
@@ -950,6 +1057,67 @@ private T PutValue<T>(TransactionOperationContext context, string type, Blittabl
             }
         }
 
+        private void PutCertificate(TransactionOperationContext context, string type, BlittableJsonReaderObject cmd, long index, ServerStore serverStore)
+        {
+            try
+            {
+                var certs = context.Transaction.InnerTransaction.OpenTable(CertificatesSchema, CertificatesSlice);
+                var command = (PutCertificateCommand)CommandBase.CreateFrom(cmd);
+
+                using (Slice.From(context.Allocator, command.PublicKeyPinningHash, out Slice hashSlice))
+                using (Slice.From(context.Allocator, command.Name, out Slice keySlice))
+                using (var cert = context.ReadObject(command.ValueToJson(), "inner-val"))
+                {
+                    var existing = serverStore.Cluster.GetCertificateByThumbprint(context, command.Name) ??
+                                  serverStore.Cluster.GetLocalStateByThumbprint(context, command.Name);
+
+                    // Ignore repeated registration of certificates
+                    if (existing != null)
+                        return;
+
+                    if (_clusterAuditLog.IsInfoEnabled)
+                        _clusterAuditLog.Info($"Registering new certificate '{command.Value.Thumbprint}' in the cluster. Security Clearance: {command.Value.SecurityClearance}. " +
+                                              $"Permissions:{Environment.NewLine}{string.Join(Environment.NewLine, command.Value.Permissions.Select(kvp => kvp.Key + ": " + kvp.Value.ToString()))}");
+
+                    UpdateCertificate(certs, keySlice, hashSlice, cert);
+                    return;
+                }
+            }
+            finally
+            {
+                NotifyValueChanged(context, type, index);
+            }
+        }
+
+        private void DiscardLeftoverCertsWithSamePinningHash(TransactionOperationContext context, string hash, string type, long index)
+        {
+            var certsWithSameHash = GetCertificatesByPinningHashSortedByExpiration(context, hash);
+
+            var keysToDelete = certsWithSameHash.Select(x => x.Thumbprint).Skip(Constants.Certificates.MaxNumberOfCertsWithSameHash).ToList();
+
+            if (keysToDelete.Count == 0)
+                return;
+
+            try
+            {
+                var certs = context.Transaction.InnerTransaction.OpenTable(CertificatesSchema, CertificatesSlice);
+
+                foreach (var key in keysToDelete)
+                {
+                    using (Slice.From(context.Allocator, key.ToLowerInvariant(), out Slice lowerKey))
```

Comment:
why aren't you calling `DeleteCertificate` here?

Reply:
fixed